db/icat.go

```diff
@@ -137,3 +216,98 @@ func (i *ICATDatabase) UserCurrentDataUsage(context context.Context, username st
 
 	return usage, nil
 }
+
+func (i *ICATDatabase) BatchCurrentDataUsage(context context.Context, start, end string) (map[string]int64, error) {
+	// Again, this should be a Tx
+	rv := make(map[string]int64)
+	s := i.UnqualifiedUsername(start)
+	e := i.UnqualifiedUsername(end)
+
+	err := i.createStorageRootMapping(context)
+	if err != nil {
+		return rv, err
+	}
+
+	resourceQuery, resourceArgs, err := i.resourcesSubselect()
+	if err != nil {
+		return rv, err
+	}
+
+	userCollsTable, err := i.createBatchUserColls(context, start, end)
+	if err != nil {
+		return rv, err
+	}
+
+	querys, args, err := i.baseUsageQuery(userCollsTable, resourceQuery, resourceArgs).
+		Where("u.user_name BETWEEN ? AND ?", s, e).
+		Column("u.user_name AS username").
+		ToSql()
+	if err != nil {
+		return rv, errors.Wrap(err, "Error formatting query")
+	}
+
+	log.Tracef("BatchCurrentDataUsage SQL: %s, %+v", querys, args)
+
+	rows, err := i.db.QueryxContext(context, querys, args...)
+	if err != nil {
+		return rv, errors.Wrap(err, "Error fetching usage for users")
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var uname string
+		var total int64
+		err = rows.Scan(&total, &uname)
+		if err != nil {
+			return rv, err
+		}
+		log.Tracef("Got %d for %s", total, uname)
+		rv[uname] = total
+	}
+	return rv, nil
+}
+
+func (i *ICATDatabase) GetUserBatchBounds(context context.Context, batchSize int) ([][]string, error) {
+	var bounds [][]string
+	// with users as (select row_number() over (order by user_name) AS n, r_user_main.user_name from r_user_main where user_type_name = 'rodsuser') select n, user_name from users where n % 100 = 0 or (n-1) % 100 = 0 union all select max(n), max(user_name) from users;
+	prefix := "WITH users AS (SELECT row_number() OVER (ORDER BY user_name) AS n, user_name FROM r_user_main WHERE user_type_name = 'rodsuser')"
+	querys, args, err := psql.Select("n", "user_name").
+		From("users").
+		Where("n % ? = 0 OR (n - 1) % ? = 0", batchSize, batchSize).
+		Prefix(prefix).
+		Suffix("UNION ALL SELECT max(n), max(user_name) FROM users").
+		ToSql()
+
+	if err != nil {
+		return bounds, errors.Wrap(err, "Error formatting SQL query")
+	}
+
+	rows, err := i.db.QueryxContext(context, querys, args...)
+	if err != nil {
+		return bounds, errors.Wrap(err, "Error fetching user batch bounds")
+	}
+	defer rows.Close()
+
+	boundsMap := make(map[int]string)
+	maxN := 0
+	for rows.Next() {
+		var rown int
+		var username string
+		err = rows.Scan(&rown, &username)
+		if err != nil {
+			return bounds, err
+		}
+		boundsMap[rown] = username
+		maxN = rown
+	}
+	log.Tracef("%d %+v", maxN, boundsMap)
+
+	for i := 1; i < maxN; i = i + batchSize {
```

Comment:
In effect, this skips over the entries for the ends of the batches (the next couple lines pull out the ends). So it's basically just a loop over each batch that should be created.